### PR TITLE
feat(search): search panel

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR adds a search panel via a webview.

image

PR checklist:

- [ ] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: src/utilities/getNonce.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * A helper function that returns a unique alphanumeric identifier called a nonce.
+ *
+ * @remarks This function is primarily used to help enforce content security
+ * policies for resources/scripts being executed in a webview context.
+ *
+ * @returns A nonce
+ */
+export function getNonce(): string {
```
Comment (ajbt200128) [inline src/utilities/getNonce.ts:9]: maybe just use something like https://github.com/uuidjs/uuid instead of writing your own nonce

File: src/views/search.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+  constructor(private readonly _extensionUri: vscode.Uri) {}
+
+  handleMessage(data: webkitCommand): void {
+    switch (data.command) {
+      case "webkit/semgrep/hello":
+        {
+          vscode.window.showInformationMessage("hello!");
+        }
+        break;
```
Comment (ajbt200128) [inline src/views/search.ts:22]: still needed?

File: src/views/search.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      // Allow scripts in the webview
+      enableScripts: true,
+
+      localResourceRoots: [this._extensionUri],
+    };
+
+    webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);
+
+    webviewView.webview.onDidReceiveMessage((data) => {
+      console.debug("Did receive message", data);
```
Comment (ajbt200128) [inline src/views/search.ts:53]: same here, still needed?

File: src/views/search.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  public resolveWebviewView(
+    webviewView: vscode.WebviewView,
+    context: vscode.WebviewViewResolveContext,
+    _token: vscode.CancellationToken
+  ): void {
+    this._view = webviewView;
+
+    webviewView.webview.options = {
+      // Allow scripts in the webview
+      enableScripts: true,
```
Comment (ajbt200128) [inline src/views/search.ts:45]: would be nice if the comment said why we want this

File: src/views/search.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    const scriptUri = webview.asWebviewUri(
+      vscode.Uri.joinPath(
+        this._extensionUri,
+        "src",
+        "webview-ui",
+        "build",
+        "assets",
+        "index.js"
+      )
+    );
```
Comment (ajbt200128) [inline src/views/search.ts:80]: would be nice if these were set previously and you only did Uri.joinpath(assets_folder, "index.js") or smthn

File: src/webview-ui/index.html [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (ajbt200128) [inline src/webview-ui/index.html:1]: is this file still needed?

File: src/webview-ui/vite.config.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { defineConfig } from "vite";
+import react from "@vitejs/plugin-react";
+
+// https://vitejs.dev/config/
+export default defineConfig({
```
Comment (ajbt200128) [inline src/webview-ui/vite.config.ts:5]: why vite? We're already using esbuild (default for vscode extensions), would prefer to use just that again if possible

File: src/searchResultsTree.ts [hunk 8]
```diff
@@ -149,7 +149,7 @@ export class SemgrepSearchProvider
     editSearch.command = { title: "Edit Search", command: "semgrep.search" };
     searchInfo.push(editSearch);
 
-    this.items = [...searchInfo, ...this.items];
+    this.items = [...this.items];
```
Comment (ajbt200128) [inline src/searchResultsTree.ts:152]: why no searchInfo here anymore? If this breaks the existing command that's ok, but let's clean that one up too if it does

**Review by ajbt200128 — APPROVED** [review]
Looks good, left some nits. This looks better than what I did, not sure if there's any dead code left over from my version but if it is let's just remove it

File: src/webview-ui/index.html (hunk 6 above)
Comment (brandonspark) [reply src/webview-ui/index.html:1]: it seems that the webview fails if this is not here

File: src/searchResultsTree.ts (hunk 8 above)
Comment (brandonspark) [reply src/searchResultsTree.ts:152]: this won't break the command, it just makes its o the icons don't appear (in favor of the different buttons we will insert)

File: src/webview-ui/README.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# `webview-ui` Directory
```
[the remark this answers is not in the document]
Comment (brandonspark) [reply src/webview-ui/README.md:1]: The webview makes things more convenient, because many aspects of the UI we want are not actually doable via the VS Code API (or, at the very least, they are quite difficult to do). Here, we have an inner react app which will later be displayed instead of a VS Code UI, so that we can have more customization over what happens.

File: src/webview-ui/vite.config.ts (hunk 7 above)
Comment (brandonspark) [reply src/webview-ui/vite.config.ts]: killed it by removing the inner module

File: src/views/search.ts (hunk 4 above)
Comment (brandonspark) [reply src/views/search.ts:39]: i don't actually know why we would want this, ill just disable it

Comment (brandonspark) [reply src/views/search.ts]: actualy it doesn't work unless you have this here
